### PR TITLE
feat: add option to retrieve offsets from tokenizer

### PR DESCRIPTION
**Pull request by riccardopinosio**
Hi,

I added the possibility to return the Offsets as a tokenizer option, could you take a look?

We use offsets in https://github.com/knights-analytics/hugot for the token classification pipeline.

### CODE REVIEW

File: tokenizer.go [hunk 1]
```diff
@@ -101,6 +105,15 @@ func uintVecToSlice(arrPtr *C.uint, len int) []uint32 {
 	return slice
 }
 
+func offsetVecToSlice(arrPtr *C.struct_Tuple, len int) []Offset {
+	arr := unsafe.Slice(arrPtr, len)
```
Comment (daulet) [inline tokenizer.go:109]: who is responsible for releasing this native struct?

File: tokenizer_test.go [hunk 2]
```diff
@@ -121,6 +128,7 @@ func TestEncodeWithAndWithoutOptions(t *testing.T) {
 			wantAttentionMask:     []uint32{0x1, 0x1},
 			wantIDs:               []uint32{101, 102},
 			wantTokens:            []string{"[CLS]", "[SEP]"},
+			wantOffsets:           []tokenizers.Offset{{0x0, 0x0}, {0x0, 0x0}},
```
Comment (daulet) [inline tokenizer_test.go:131]: i'd expect this slice to be empty when offsets are not requested

File: tokenizer.go (hunk 1 above)
Comment (riccardopinosio) [reply tokenizer.go]: @daulet indeed that was a leak, I fixed it by doing things differently, i.e., avoiding the tuple struct thing alltogether and just flattening the offset vector returned by the Rust tokenizer, passing that to the go side like the other vectors (only difference is this vector is of length 2*len), and recombining it on the go side to a slice of offsets. See last commit.

File: tokenizer_test.go (hunk 2 above)
Comment (riccardopinosio) [reply tokenizer_test.go:131]: I think this test requests all options, but there are tests below where the offset is not requested and the returned slice is indeed empty.

File: tokenizer.go [hunk 3]
```diff
@@ -201,6 +224,10 @@ func (t *Tokenizer) EncodeWithOptions(str string, addSpecialTokens bool, opts ..
 		encoding.AttentionMask = uintVecToSlice(res.attention_mask, len)
 	}
 
+	if encOptions.ReturnOffsets && res.offsets != nil {
+		encoding.Offsets = offsetVecToSlice(res.offsets, len)
```
Comment (daulet) [inline tokenizer.go:228]: i believe you still need to free `res.offsets`

**Review by daulet — APPROVED** [review]

File: tokenizer.go (hunk 3 above)
Comment (riccardopinosio) [reply tokenizer.go:228]: I added a line to free the offsets inside the if statement.

Comment (daulet) [reply tokenizer.go:228]: SGTM

**Comment by daulet** [comment]
@riccardopinosio PTAL at the test failure in CI

**Comment by riccardopinosio** [comment]
@daulet I reverted the change that added the freeing of the offset array memory, bazel didn't like it because memory was being freed twice. I believe the offsets get freed when the buffer itself is freed using `C.free_buffer`

**Review by daulet — APPROVED** [review]

**Pull request merged**